### PR TITLE
Fix crash issue of TelemetryBidiReactor and MetricBidiReactor

### LINKED ISSUE #789: [Bug] 5.x版本cpp客户端，同时producer和push_consumer，持续运行，crashed1
### Before Creating the Bug Report

- [X] I found a bug, not just asking a question, which should be created in [GitHub Discussions](https://github.com/apache/rocketmq-clients/discussions).

- [X] I have searched the [GitHub Issues](https://github.com/apache/rocketmq-clients/issues) and [GitHub Discussions](https://github.com/apache/rocketmq-clients/discussions)  of this repository and believe that this is not a duplicate.

- [X] I have confirmed that this bug belongs to the current repository, not other repositories of RocketMQ.

### Programming Language of the Client

C++

### Runtime Platform Environment

Red Hat Enterprise Linux Server release 7.9 (Maipo)

### RocketMQ Version of the Client/Server

客户端：2024.07.05拉取master分支代码
服务端：rocketmq 5.1.4

### Run or Compiler Version

Red Hat Enterprise Linux Server release 7.9 (Maipo)

### Describe the Bug

Program terminated with signal 6, Aborted.
#0  0x00007ff7604e4387 in raise () from TestRocketMQModule-runtime/libs/libc.so.6
Missing separate debuginfos, use: debuginfo-install glibc-2.17-326.0.1.el7_9.x86_64
(gdb) bt
#0  0x00007ff7604e4387 in raise () from TestRocketMQModule-runtime/libs/libc.so.6
#1  0x00007ff7604e5a78 in abort () from TestRocketMQModule-runtime/libs/libc.so.6
#2  0x00007ff74752465a in absl::lts_20211102::raw_logging_internal::(anonymous namespace)::RawLogVA(absl::lts_20211102::LogSeverity, char const*, int, char const*, __va_list_tag*) () from TestRocketMQModule-runtime/libs/librocketmq.so
#3  0x00007ff7475247a0 in absl::lts_20211102::raw_logging_internal::RawLog(absl::lts_20211102::LogSeverity, char const*, int, char const*, ...) () from TestRocketMQModule-runtime/libs/librocketmq.so
#4  0x00007ff7474c8bb7 in absl::lts_20211102::CheckForMutexCorruption(long, char const*) ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#5  0x00007ff7474c8d50 in absl::lts_20211102::Mutex::LockSlowLoop(absl::lts_20211102::SynchWaitParams*, int) ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#6  0x00007ff7474c8a75 in absl::lts_20211102::Mutex::LockSlowWithDeadline(absl::lts_20211102::MuHowS const*, absl::lts_20211102::Condition const*, absl::lts_20211102::synchronization_internal::KernelTimeout, int) ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#7  0x00007ff7474c8752 in absl::lts_20211102::Mutex::LockSlow(absl::lts_20211102::MuHowS const*, absl::lts_20211102::Condition const*, int) () from TestRocketMQModule-runtime/libs/librocketmq.so
#8  0x00007ff7474c7948 in absl::lts_20211102::Mutex::Lock() () from TestRocketMQModule-runtime/libs/librocketmq.so
#9  0x00007ff746bd5128 in absl::lts_20211102::MutexLock::MutexLock(absl::lts_20211102::Mutex*) ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#10 0x00007ff746c2a849 in rocketmq::TelemetryBidiReactor::tryWriteNext() ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#11 0x00007ff746c28362 in rocketmq::TelemetryBidiReactor::OnWriteDone(bool) ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#12 0x00007ff746d92b10 in _ZZN4grpc8internal30ClientCallbackReaderWriterImplIN6apache8rocketmq2v216TelemetryCommandES5_EC4ENS0_4CallEPNS_13ClientContextEPNS_17ClientBidiReactorIS5_S5_EEENKUlbE0_clEb ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#13 0x00007ff746d95a82 in _ZNSt17_Function_handlerIFvbEZN4grpc8internal30ClientCallbackReaderWriterImplIN6apache8rocketmq2v216TelemetryCommandES7_EC4ENS2_4CallEPNS1_13ClientContextEPNS1_17ClientBidiReactorIS7_S7_EEEUlbE0_E9_M_invokeERKSt9_Any_dataOb
    () from TestRocketMQModule-runtime/libs/librocketmq.so
#14 0x00007ff746d1dab6 in std::function<void (bool)>::operator()(bool) const ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#15 0x00007ff746d1b7c9 in void grpc::internal::CatchingCallback<std::function<void (bool)>&, bool&>(std::function<void (bool)>&, bool&) () from TestRocketMQModule-runtime/libs/librocketmq.so
#16 0x00007ff746d1949a in grpc::internal::CallbackWithSuccessTag::Run(bool) ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#17 0x00007ff746d193e0 in grpc::internal::CallbackWithSuccessTag::StaticRun(grpc_completion_queue_functor*, int) ()
   from TestRocketMQModule-runtime/libs/librocketmq.so
#18 0x00007ff746cfc457 in grpc::(anonymous namespace)::CallbackAlternativeCQ::Ref()::{lambda(void*)#1}::operator()(void*) const () from TestRocketMQModule-runtime/libs/librocketmq.so
#19 0x00007ff746cfc485 in grpc::(anonymous namespace)::CallbackAlternativeCQ::Ref()::{lambda(void*)#1}::_FUN(void*) ()
---Type <return> to continue, or q <return> to quit---
   from TestRocketMQModule-runtime/libs/librocketmq.so
#20 0x00007ff74747a7ed in _ZZN9grpc_core12_GLOBAL__N_120ThreadInternalsPosixC4EPKcPFvPvES4_PbRKNS_6Thread7OptionsEENKUlS4_E_clES4_ () from TestRocketMQModule-runtime/libs/librocketmq.so
#21 0x00007ff74747a81e in _ZZN9grpc_core12_GLOBAL__N_120ThreadInternalsPosixC4EPKcPFvPvES4_PbRKNS_6Thread7OptionsEENUlS4_E_4_FUNES4_ () from TestRocketMQModule-runtime/libs/librocketmq.so
#22 0x00007ff760883ea5 in start_thread () from TestRocketMQModule-runtime/libs/libpthread.so.0
#23 0x00007ff7605acb0d in clone () from TestRocketMQModule-runtime/libs/libc.so.6

### Steps to Reproduce

5.x版本cpp客户端，同时运行producer和push_consumer，持续运行

### What Did You Expect to See?

1

### What Did You See Instead?

1

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by 0902horn**
### Which Issue(s) This PR Fixes

Fixes #789 #790 #791 #792 #795 #796, and maybe #794 #797 .

### CODE REVIEW

File: cpp/source/client/TelemetryBidiReactor.cpp [hunk 1]
```diff
@@ -383,134 +279,61 @@ void TelemetryBidiReactor::write(TelemetryCommand command) {
...
-void TelemetryBidiReactor::fireClose() {
+void TelemetryBidiReactor::signalClose() {
+  absl::MutexLock lk(&state_mtx_);
+  state_ = StreamState::Closing;
+}
+
+void TelemetryBidiReactor::close() {
   SPDLOG_INFO("{}#fireClose", peer_address_);
 
+  if (peer_address_.rfind(":8080") != std::string::npos) {
```
Comment (lizhanhui) [inline cpp/source/client/TelemetryBidiReactor.cpp:308]: Why we need this?

**Comment by lizhanhui** [comment]
grpc::ClientContext#TryCancel is something I missed and It helps simplify the workflow a lot.  Good Job.

File: cpp/source/client/include/TelemetryBidiReactor.h [hunk 2]
```diff
@@ -122,7 +111,7 @@ class TelemetryBidiReactor : public grpc::ClientBidiReactor<TelemetryCommand, Te
   TelemetryCommand read_;
 
   /**
-   * @brief Buffered commands to write to server
+   * @brief Buffered commands to write to serverd
```
Comment (lizhanhui) [inline cpp/source/client/include/TelemetryBidiReactor.h:114]: Typo, drop 'd'

File: cpp/source/stats/include/MetricBidiReactor.h [hunk 3]
```diff
@@ -36,8 +36,8 @@ class MetricBidiReactor
     : public grpc::ClientBidiReactor<ExportMetricsServiceRequest,
                                      ExportMetricsServiceResponse> {
 public:
-  MetricBidiReactor(std::weak_ptr<Client> client,
-                    std::weak_ptr<OpencensusExporter> exporter);
+  MetricBidiReactor(std::shared_ptr<Client> client,
```
Comment (lizhanhui) [inline cpp/source/stats/include/MetricBidiReactor.h:39]: Why shared? weak_ptr shall suffice.

File: cpp/source/client/TelemetryBidiReactor.cpp (hunk 1 above)
Comment (0902horn) [reply cpp/source/client/TelemetryBidiReactor.cpp:308]: This line was for debugging  and should removed.

File: cpp/source/stats/include/MetricBidiReactor.h (hunk 3 above)
Comment (0902horn) [reply cpp/source/stats/include/MetricBidiReactor.h:39]: This is to fix the crash happened in MetricBidiReactor.

File: cpp/source/client/include/TelemetryBidiReactor.h [hunk 4]
```diff
@@ -97,21 +86,21 @@ class TelemetryBidiReactor : public grpc::ClientBidiReactor<TelemetryCommand, Te
...
-  ///               further Start* should be called.
-  void OnWritesDoneDone(bool ok) override;
+//  /// Notifies the application that a StartWritesDone operation completed. Note
+//  /// that this is only used on explicit StartWritesDone operations and not for
+//  /// those that are implicitly invoked as part of a StartWriteLast.
+//  ///
+//  /// \param[in] ok Was it successful? If false, the application will later see
+//  ///               the failure reflected as a bad status in OnDone and no
+//  ///               further Start* should be called.
+//  void OnWritesDoneDone(bool ok) override;
```
Comment (lizhanhui) [inline cpp/source/client/include/TelemetryBidiReactor.h:96]: We may remove it as it's no longer used.

**Review by lizhanhui — APPROVED** [review]
OK

**Review by lizhanhui — APPROVED** [review]
OK

**Pull request merged**